crates/viewer/re_ui/src/filter_widget.rs

```diff
@@ -0,0 +1,276 @@
+        let mut button_clicked = false;
+
+        let icon = if self.inner_state.is_none() {
+            &crate::icons::SEARCH
+        } else {
+            &crate::icons::CLOSE
+        };
+
+        // precompute the title layout such that we know the size we need for the list item content
+        let section_title = section_title.into();
+        let galley = section_title.into_galley(
+            ui,
+            Some(egui::TextWrapMode::Extend),
+            f32::INFINITY,
+            egui::FontSelection::default(),
+        );
+        let text_width = galley.size().x;
+
+        list_item::list_item_scope(ui, ui.next_auto_id(), |ui| {
+            ui.list_item()
+                .interactive(false)
+                .with_height(30.0)
+                .show_flat(
+                    ui,
+                    list_item::CustomContent::new(|ui, _| {
+                        if self.inner_state.is_some()
+                            && ui.input_mut(|i| {
+                                i.consume_key(egui::Modifiers::NONE, egui::Key::Escape)
+                            })
+                        {
+                            self.inner_state = None;
+                        }
+
+                        if let Some(inner_state) = self.inner_state.as_mut() {
+                            // we add additional spacing for aesthetic reasons (active text edits have a
+                            // fat border)
+                            ui.spacing_mut().text_edit_width =
+                                (ui.max_rect().width() - 10.0).at_least(0.0);
+
+                            let response =
+                                egui::TextEdit::singleline(&mut inner_state.filter_query)
+                                    .lock_focus(true)
+                                    .ui(ui);
+
+                            if self.request_focus {
+                                self.request_focus = false;
+                                response.request_focus();
+                            }
+                        } else {
+                            ui.label(galley);
+                        }
+                    })
+                    .with_content_width(text_width)
+                    .action_button(icon, || {
+                        button_clicked = true;
+                    }),
+                );
+        });
+
+        // defer button handling because we can't mutably borrow `self` in both closures above
+        if button_clicked {
+            if self.inner_state.is_none() {
+                self.inner_state = Some(InnerState::default());
+                self.request_focus = true;
+            } else {
+                self.inner_state = None;
+            }
+        }
+    }
+}
+
+// --
+
+/// Full-text, case-insensitive matcher.
+pub struct FilterMatcher {
+    /// The lowercase version of the query string.
+    ///
+    /// If this is `None` or `Some("")`, the matcher will accept any input.
```

Comment:
```suggestion
    /// If this is `None` the matcher will accept any input.
    /// If this is `Some("")` the matcher will reject any input.
```

That's at least what's consistently implemented down there!